## Stock Data
Now that you've had exposure to time series data, let's look at bringing stock prices into Pandas.
## Reading in Data
Your dataset can come in a variety of different formats. The most common format is the [CSV](https://en.wikipedia.org/wiki/Comma-separated_values). We'll use the "prices.csv" file as an example csv file.

In [144]:
# The code was removed by Watson Studio for sharing.

ticker        date         open         high          low        close  \
0    ABC  2017-09-05 163.09000000 164.24000000 160.21000000 162.63000000   
1    ABC  2017-09-06 162.85000000 162.46000000 159.99000000 161.13000000   
2    ABC  2017-09-07 162.11000000 162.70000000 160.65000000 161.26000000   
3    ABC  2017-09-08 160.41000000 160.89000000 159.31000000 158.05000000   
4    ABC  2017-09-11 161.09000000 162.14000000 159.54000000 161.29000000   

             volume    adj_close        adj_volume  
0 29417590.00000000 162.49000000 29414672.00000000  
1 21131267.00000000 162.44000000 21169319.00000000  
2 21722502.00000000 161.46000000 21719856.00000000  
3 28311012.00000000 158.26000000 28305810.00000000  
4 31075573.00000000 160.97000000 31163734.00000000

In [145]:

body = client_a62c8cfdc9c54b598152b0aa1925b762.get_object(Bucket='aiprojects-donotdelete-pr-vpy9lukvrmyy6w',Key='prices_2017_09_22_2017-09-28.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_2 = pd.read_csv(body,names=['ticker', 'date', 'open', 'high', 'low',
                                             'close', 'volume', 'adj_close', 'adj_volume'])
df_data_2.head()



ticker        date         open         high          low        close  \
0      A  2017-09-22 152.46000000 152.72000000 150.31000000 152.48000000   
1      A  2017-09-25 149.54000000 151.93000000 148.46000000 151.11000000   
2      A  2017-09-26 152.33000000 154.31000000 152.41000000 152.42000000   
3      A  2017-09-27 153.43000000 155.36000000 152.86000000 154.34000000   
4      A  2017-09-28 153.46000000 154.90000000 152.34000000 153.68000000   

             volume    adj_close        adj_volume  
0 45904657.00000000 151.78000000 46179797.00000000  
1 43886111.00000000 151.11000000 43724099.00000000  
2 35431266.00000000 153.10000000 35345230.00000000  
3 25027198.00000000 154.31000000 24880751.00000000  
4 21928819.00000000 153.34000000 21995499.00000000

In [146]:
#with open('prices.csv', 'r') as file:
#    prices = file.read()
    
#print(prices)

The data provider will provide you with information for each field in the CSV. This csv has the fields ticker, date, open, high, low, close, volume, adj_close, adj_volume in that order. That means, the first line in the CSV has the following data:

- ticker: ABC
- date: 2017-09-05
- open: 163.09
- high: 164.24
- low: 160.21
- close: 162.63
- volume: 29417590.0
- adj_close: 162.49
- adj_volume: 29414672.0

Let's move this data into a DataFrame. For this, we'll need to use the [`pd.read_csv`](https://pandas.pydata.org/pandas-docs/version/0.21.1/generated/pandas.read_csv.html) function. This allows you generate a DataFrame from CSV data.

In [147]:
import pandas as pd

#price_df = df_data_1

#price_df

That generated a DataFrame using the CSV, but assumed the first row contains the field names. We'll have to supply the function's parameter `names` with a list of fiels names.

In [148]:
#price_df = pd.read_csv(body, names=['ticker', 'date', 'open', 'high', 'low',
#                                             'close', 'volume', 'adj_close', 'adj_volume'])
price_df=df_data_1.copy()

price_df

ticker        date         open         high          low        close  \
0     ABC  2017-09-05 163.09000000 164.24000000 160.21000000 162.63000000   
1     ABC  2017-09-06 162.85000000 162.46000000 159.99000000 161.13000000   
2     ABC  2017-09-07 162.11000000 162.70000000 160.65000000 161.26000000   
3     ABC  2017-09-08 160.41000000 160.89000000 159.31000000 158.05000000   
4     ABC  2017-09-11 161.09000000 162.14000000 159.54000000 161.29000000   
5     ABC  2017-09-12 162.54000000 164.61000000 159.52000000 161.09000000   
6     ABC  2017-09-13 160.01000000 160.51000000 158.22000000 159.29000000   
7     EFG  2017-09-05 154.45000000 154.69000000 153.17000000 154.52000000   
8     EFG  2017-09-06 155.03000000 155.14000000 153.89000000 154.45000000   
9     EFG  2017-09-07 154.73000000 155.36000000 153.60000000 155.68000000   
10    EFG  2017-09-08 156.01000000 155.91000000 154.17000000 155.86000000   
11    EFG  2017-09-11 157.07000000 157.71000000 155.93000000 157.17000000   
12    EFG  2017-09-12 155.98000000 156.72000000 154.28000000 156.71000000   
13    EFG  2017-09-13 156.40000000 157.07000000 155.68000000 155.54000000   
14    XYZ  2017-09-05  63.90000000  64.51000000  63.13000000  63.95000000   
15    XYZ  2017-09-06  63.85000000  63.65000000  61.72000000  62.23000000   
16    XYZ  2017-09-07  61.97000000  61.93000000  59.47000000  60.46000000   
17    XYZ  2017-09-08  60.36000000  60.45000000  58.51000000  59.35000000   
18    XYZ  2017-09-11  60.04000000  59.92000000  57.68000000  58.24000000   
19    XYZ  2017-09-12  58.19000000  59.29000000  57.89000000  58.71000000   
20    XYZ  2017-09-13  59.01000000  60.66000000  58.80000000  60.33000000   

              volume    adj_close        adj_volume  
0  29417590.00000000 162.49000000 29414672.00000000  
1  21131267.00000000 162.44000000 21169319.00000000  
2  21722502.00000000 161.46000000 21719856.00000000  
3  28311012.00000000 158.26000000 28305810.00000000  
4  31075573.00000000 160.97000000 31163734.00000000  
5  70921229.00000000 160.62000000 71097150.00000000  
6  44580353.00000000 159.07000000 44260255.00000000  
7   1270203.00000000 153.58000000  1270679.00000000  
8   1195987.00000000 154.06000000  1196107.00000000  
9   1420730.00000000 155.60000000  1409098.00000000  
10  1438929.00000000 156.08000000  1445338.00000000  
11  1608840.00000000 156.70000000  1610357.00000000  
12  1692197.00000000 156.86000000  1687819.00000000  
13  1211779.00000000 155.85000000  1210716.00000000  
14  1738651.00000000  63.33000000  1733249.00000000  
15  3730110.00000000  61.95000000  3725435.00000000  
16  6166046.00000000  60.64000000  6191712.00000000  
17  5173590.00000000  59.40000000  5174940.00000000  
18  5003322.00000000  58.02000000  5001118.00000000  
19  3633446.00000000  58.96000000  3635132.00000000  
20  3571591.00000000  60.46000000  3583560.00000000

## DataFrame Calculations
Now that we have the data in a DataFrame, we can start to do calculations on it. Let's find out the median value for each stock using the [`DataFrame.median`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.median.html) function.

In [149]:
price_df.median()

open             155.98000000
high             155.91000000
low              154.17000000
close            155.68000000
volume       3730110.00000000
adj_close        155.85000000
adj_volume   3725435.00000000
dtype: float64

That's not right. Those are the median values for the whole stock universe. We'll use the [`DataFrame.groupby`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.groupby.html) function to get mean for each stock.

In [150]:
price_df.groupby('ticker').median()

open         high          low        close            volume  \
ticker                                                                         
ABC    162.11000000 162.46000000 159.54000000 161.13000000 29417590.00000000   
EFG    155.98000000 155.91000000 154.17000000 155.68000000  1420730.00000000   
XYZ     60.36000000  60.66000000  58.80000000  60.33000000  3730110.00000000   

          adj_close        adj_volume  
ticker                                 
ABC    160.97000000 29414672.00000000  
EFG    155.85000000  1409098.00000000  
XYZ     60.46000000  3725435.00000000

That's what we're looking for! However, we don't want to run the `groupby` function each time we make an operation. We could save the GroupBy object by doing `price_df_ticker_groups = price_df.groupby('ticker')`. This limits us to the operations of GroupBy objects. There's the [`GroupBy.apply`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.core.groupby.GroupBy.apply.html), but then we lose out on performance. The true problem is the way the data is represented.

In [151]:
price_df.iloc[:16]

ticker        date         open         high          low        close  \
0     ABC  2017-09-05 163.09000000 164.24000000 160.21000000 162.63000000   
1     ABC  2017-09-06 162.85000000 162.46000000 159.99000000 161.13000000   
2     ABC  2017-09-07 162.11000000 162.70000000 160.65000000 161.26000000   
3     ABC  2017-09-08 160.41000000 160.89000000 159.31000000 158.05000000   
4     ABC  2017-09-11 161.09000000 162.14000000 159.54000000 161.29000000   
5     ABC  2017-09-12 162.54000000 164.61000000 159.52000000 161.09000000   
6     ABC  2017-09-13 160.01000000 160.51000000 158.22000000 159.29000000   
7     EFG  2017-09-05 154.45000000 154.69000000 153.17000000 154.52000000   
8     EFG  2017-09-06 155.03000000 155.14000000 153.89000000 154.45000000   
9     EFG  2017-09-07 154.73000000 155.36000000 153.60000000 155.68000000   
10    EFG  2017-09-08 156.01000000 155.91000000 154.17000000 155.86000000   
11    EFG  2017-09-11 157.07000000 157.71000000 155.93000000 157.17000000   
12    EFG  2017-09-12 155.98000000 156.72000000 154.28000000 156.71000000   
13    EFG  2017-09-13 156.40000000 157.07000000 155.68000000 155.54000000   
14    XYZ  2017-09-05  63.90000000  64.51000000  63.13000000  63.95000000   
15    XYZ  2017-09-06  63.85000000  63.65000000  61.72000000  62.23000000   

              volume    adj_close        adj_volume  
0  29417590.00000000 162.49000000 29414672.00000000  
1  21131267.00000000 162.44000000 21169319.00000000  
2  21722502.00000000 161.46000000 21719856.00000000  
3  28311012.00000000 158.26000000 28305810.00000000  
4  31075573.00000000 160.97000000 31163734.00000000  
5  70921229.00000000 160.62000000 71097150.00000000  
6  44580353.00000000 159.07000000 44260255.00000000  
7   1270203.00000000 153.58000000  1270679.00000000  
8   1195987.00000000 154.06000000  1196107.00000000  
9   1420730.00000000 155.60000000  1409098.00000000  
10  1438929.00000000 156.08000000  1445338.00000000  
11  1608840.00000000 156.70000000  1610357.00000000  
12  1692197.00000000 156.86000000  1687819.00000000  
13  1211779.00000000 155.85000000  1210716.00000000  
14  1738651.00000000  63.33000000  1733249.00000000  
15  3730110.00000000  61.95000000  3725435.00000000

Can you spot our problem? Take a moment to see if you can find it.

The problem is between lines [6, 7] and  [13, 14]

In [152]:
price_df.iloc[[6, 7, 13, 14]]

ticker        date         open         high          low        close  \
6     ABC  2017-09-13 160.01000000 160.51000000 158.22000000 159.29000000   
7     EFG  2017-09-05 154.45000000 154.69000000 153.17000000 154.52000000   
13    EFG  2017-09-13 156.40000000 157.07000000 155.68000000 155.54000000   
14    XYZ  2017-09-05  63.90000000  64.51000000  63.13000000  63.95000000   

              volume    adj_close        adj_volume  
6  44580353.00000000 159.07000000 44260255.00000000  
7   1270203.00000000 153.58000000  1270679.00000000  
13  1211779.00000000 155.85000000  1210716.00000000  
14  1738651.00000000  63.33000000  1733249.00000000

Data for all the tickers are stacked. We're representing 3 dimensional data in 2 dimensions. This was solved using Panda's Panels, which is [deprecated](https://pandas.pydata.org/pandas-docs/version/0.21.0/dsintro.html#deprecate-panel). The Pandas documentation recommends we use either [MultiIndex](https://pandas.pydata.org/pandas-docs/version/0.21/advanced.html) or [xarray](http://xarray.pydata.org/en/stable/). [MultiIndex](https://pandas.pydata.org/pandas-docs/version/0.21/advanced.html) still doesn't solve our problem, since the data is still represented in 2 dimensions. [xarray](http://xarray.pydata.org/en/stable/) is able to store 3 dimensional data, but Finance uses Pandas, so we'll stick with this library. After you finish this program, I recommend you check out [xarray](http://xarray.pydata.org/en/stable/).

So, how do we use our 3-dimensional data with Pandas? We can split each 3rd dimension into it's own 2 dimension DataFrame. Let's take this array as an example:
```
[
    [
        [ 0,  1],
        [ 2,  3],
        [ 4,  5]
    ],[
        [ 6,  7],
        [ 8,  9],
        [10, 11]
    ],[
        [12, 13],
        [14, 15],
        [16, 17]
    ],[
        [18, 19],
        [20, 21],
        [22, 23]
    ]
]    
```
We want to split it into these two 2d arrays:
```
[
    [0, 2, 4],
    [6, 8, 10],
    [12, 14, 16],
    [18, 20, 22]
] 
[
    [1, 3, 5],
    [7, 8, 11],
    [13, 15, 17],
    [19, 21, 23]
] 
```
In our case, our third dimensions are "open", "high", "low", "close", "volume", "adj_close", and  "adj_volume". We'll use the [`DataFrame.pivot`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.pivot.html) function to generate these DataFrames.

In [153]:
open_prices = price_df.pivot(index='date', columns='ticker', values='open')
high_prices = price_df.pivot(index='date', columns='ticker', values='high')
low_prices = price_df.pivot(index='date', columns='ticker', values='low')
close_prices = price_df.pivot(index='date', columns='ticker', values='close')
volume = price_df.pivot(index='date', columns='ticker', values='volume')
adj_close_prices = price_df.pivot(index='date', columns='ticker', values='adj_close')
adj_volume = price_df.pivot(index='date', columns='ticker', values='adj_volume')
#test_prices = price_df.pivot(index='date', columns='ticker')
open_prices
#test_prices

ticker              ABC          EFG         XYZ
date                                            
2017-09-05 163.09000000 154.45000000 63.90000000
2017-09-06 162.85000000 155.03000000 63.85000000
2017-09-07 162.11000000 154.73000000 61.97000000
2017-09-08 160.41000000 156.01000000 60.36000000
2017-09-11 161.09000000 157.07000000 60.04000000
2017-09-12 162.54000000 155.98000000 58.19000000
2017-09-13 160.01000000 156.40000000 59.01000000

That gives you DataFrames for all the open, high low, etc.. Now, what we have been waiting for.. The mean for each ticker.

In [154]:
open_prices.mean()

ticker
ABC   161.72857143
EFG   155.66714286
XYZ    61.04571429
dtype: float64

We can also get the mean for each date by doing a transpose.

In [155]:
#open_prices.T.mean()
open_prices.mean(axis=1)

date
2017-09-05   127.14666667
2017-09-06   127.24333333
2017-09-07   126.27000000
2017-09-08   125.59333333
2017-09-11   126.06666667
2017-09-12   125.57000000
2017-09-13   125.14000000
dtype: float64

It doesn't matter whether date is the index and tickers are the colums or the other way around. It's always a transpose away. Since we're going to do a lot of operations across dates, we will stick with date as the index and tickers as the colums throughtout this program.
## Quiz
Let's see if you can apply what you learned. Implment the `csv_to_pivot` function to take in a filepath, `csv_filename`, and output the close 2d array. You can assume the CSV file used by `csv_to_close` has the same field names as "prices.csv" and in the same order.

To help with your implemention of quizzes, we provide you with unit tests to test your function implemention. For this quiz, we'll be using the function `test_csv_to_close` in the `quiz_tests` module to test `csv_to_close`.

In [157]:
import quiz_tests


def csv_to_close(csv_filepath, field_names):
    """Reads in data from a csv file and produces a DataFrame with close data.
    
    Parameters
    ----------
    csv_filepath : str
        The name of the csv file to read
    field_names : list of str
        The field names of the field in the csv file

    Returns
    -------
    close : DataFrame
        Close prices for each ticker and date
    """
    
    # TODO: Implement Function
    #price_df = pd.read_csv(csv_filepath,names=field_names)
    price_df=df_data_2
    #print(price_df.head())
    close_prices = price_df.pivot(index='date', columns='ticker', values='close')

    #print(price_df.head())
    
    return close_prices


quiz_tests.test_csv_to_close(csv_to_close)


## Quiz Solution
If you're having trouble, you can check out the quiz solution [here](stock_data_solution.ipynb).